src/libraries/System.Private.CoreLib/src/System/Number.NumberToFloatingPointBits.cs

```diff
@@ -361,39 +1084,60 @@ private static ulong NumberToDoubleFloatingPointBits(ref NumberBuffer number, in
             uint integerDigitsPresent = Math.Min(positiveExponent, totalDigits);
             uint fractionalDigitsPresent = totalDigits - integerDigitsPresent;
 
-            uint fastExponent = (uint)(Math.Abs(number.Scale - integerDigitsPresent - fractionalDigitsPresent));
+            // Above 19 digits, we rely on slow path
+            if (totalDigits <= 19)
+            {
+                byte* src = number.GetDigitsPointer();
 
-            // When the number of significant digits is less than or equal to 15 and the
-            // scale is less than or equal to 22, we can take some shortcuts and just rely
-            // on floating-point arithmetic to compute the correct result. This is
-            // because each floating-point precision values allows us to exactly represent
-            // different whole integers and certain powers of 10, depending on the underlying
-            // formats exact range. Additionally, IEEE operations dictate that the result is
-            // computed to the infinitely precise result and then rounded, which means that
-            // we can rely on it to produce the correct result when both inputs are exact.
+                ulong mantissa = DigitsToUInt64(src, (int)(totalDigits));
 
-            byte* src = number.GetDigitsPointer();
+                int exponent = (int)(number.Scale - integerDigitsPresent - fractionalDigitsPresent);
+                int fastExponent = (Math.Abs(exponent));
 
-            if ((totalDigits <= 15) && (fastExponent <= 22))
-            {
-                double result = DigitsToUInt64(src, (int)(totalDigits));
-                double scale = s_Pow10DoubleTable[fastExponent];
+                // When the number of significant digits is less than or equal to MaxMantissaFastPath and the
+                // scale is less than or equal to MaxExponentFastPath, we can take some shortcuts and just rely
+                // on floating-point arithmetic to compute the correct result. This is
+                // because each floating-point precision values allows us to exactly represent
+                // different whole integers and certain powers of 10, depending on the underlying
+                // formats exact range. Additionally, IEEE operations dictate that the result is
+                // computed to the infinitely precise result and then rounded, which means that
+                // we can rely on it to produce the correct result when both inputs are exact.
+                // This is known as Clinger's fast path
 
-                if (fractionalDigitsPresent != 0)
+                if ((mantissa <= info.MaxMantissaFastPath) && (totalDigits <= 15) && (fastExponent <= info.MaxExponentFastPath))
```

Comment:
@CarlVerret You have three checks here, are the three checks needed? You check for the number of total digits and for the mantissa. It seems that there is one too many check. No?

Reply:
The `(totalDigits <= 15)` is suspicious... maybe spurious?